src/printer.js

```diff
@@ -2762,62 +2763,83 @@ function printJSXChildren(path, options, print, jsxWhitespace) {
           : util.htmlEscapeInsideAngleBracket(child.value);
         const value = partiallyEscapedValue.replace(/\u00a0/g, "&nbsp;");
 
-        if (/\S/.test(value)) {
-          // treat each line of text as its own entity
-          value.split(/(\n\s*)/).forEach(line => {
-            const newlines = line.match(/\n/g);
-            if (newlines) {
-              children.push(hardline);
-
-              // allow one extra newline
-              if (newlines.length > 1) {
-                children.push(hardline);
-              }
-              return;
-            }
 
-            const beginSpace = /^\s+/.test(line);
-            if (beginSpace) {
-              children.push(jsxWhitespace);
-              children.push(softline);
-            }
+        // // console.log('value', value);
+        // if (/\S/.test(value)) {
+        //   value.split(/(\s+)/).forEach(word => {
+        //     // const newlines = word.match(/\n/g);
+        //     // if (newlines) {
+        //     //   children.push(hardline);
+        //     //
+        //     //   // allow one extra newline
+        //     //   if (newlines.length > 1) {
+        //     //     children.push(hardline);
+        //     //   }
+        //     //   return;
+        //     // }
+        //
+        //     const space = /^\s+$/.test(word);
+        //     if (space) {
+        //       words.push(line);
+        //       return;
+        //     }
+        //
+        //     const stripped = word.replace(/^\s+|\s+$/g, "");
+        //     if (stripped) {
+        //       words.push(stripped);
+        //     }
+        //   });
+        //
+        //   if (!isLineNext(util.getLast(words))) {
+        //     words.push(softline);
+        //   }
+        // } else if (/\n/.test(value)) {
+        //   words.push(hardline);
+        //
+        //   // allow one extra newline
+        //   if (value.match(/\n/g).length > 1) {
+        //     words.push(hardline);
+        //   }
+        // } else if (/\s/.test(value)) {
+        //   // whitespace-only without newlines,
+        //   // eg; a single space separating two elements
+        //   words.push(jsxWhitespace);
+        //   words.push(softline);
+        // }
+
+        value.replace(/^\s+|\s+$/g, "").split(/(\s+)/).forEach(word => {
```

Comment:
I haven't looked at this super in-depth, but it looks like the intent here would be to squash jsx+element mixtures to be as tight as possible?

So, 

```jsx
<div>
  hello
  world
</div>
```
would become 

```jsx
<div>
  hello world
</div>
```
and 

```jsx
<div>
  hello
  <world />
</div>
```
would be come

```jsx
<div>
  hello<world />
</div>
```

?

(maybe you haven't thought this far ahead yet, which is fine – or maybe you plan to go back to the commented-out code soon enough)